File: Queue.cc
#include <iostream>
#include <string>
using namespace std;

class Que {
public:
	void enqueue(int);
	int dequeue();
	void disp();
	~Que();
private:
	class node {
	public:
		int data;
		node* next;
		node* prev;
		node(int n) : data(n) {
			prev = next = nullptr;
		};
	};
	node* head = nullptr;
	node* tail = nullptr;
	node* cur = nullptr; // current
	node* pre = nullptr; // prev
	int ncnt = 0; // node counter
};

void Que::enqueue(int n) {
	if (!ncnt) {
		head = new node(n);
		tail = head;
	}
	else if (ncnt == 1) {
		tail = new node(n);
		head->next = tail;
		tail->prev = head;
	}
	else {
		cur = tail;
		tail = new node(n);
		cur->next = tail;
		tail->prev = cur;
	}
	ncnt++;
}
int Que::dequeue() {
	if (ncnt < 1) {
		cout << "Que is empty";
		return -1;
	}
	int tmp = head->data;
	if (ncnt == 1) {
		delete head;
	}
	else if (ncnt == 2) {
		delete head;
		head = tail;
	}
	else {
		cur = head->next;
		delete head;
		head = cur;
	}
	ncnt--;
	return tmp;
}

void Que::disp() {
	string msg;
	if (ncnt < 1) {
		msg = "data does not exist\n\n";
	}
	else {
		msg = "list: [ ";
		if (head == tail) {
			msg += to_string(head->data);
		}

		else {
			cur = head;
			for (int i = 0; cur; i++) {
				msg += to_string(cur->data);
				if (i < ncnt - 1) msg += ", ";
				cur = cur->next;
			}
		}
		msg += " ]\n\n";
	}
	cout << msg;
}

Que::~Que() {
	if (ncnt == 1) {
		delete head;
		return;
	}
	for (int i = 0; i < ncnt; i++) {
		if (!head->next) {
			delete head;
			return;
		}
		cur = head->next;
		delete head;
		head = cur;
	}
}

int main() {
	Que que;
	que.enqueue(3);
	que.enqueue(2);
	que.disp();
	que.dequeue();
	que.disp();
}

File: Queue_cs/Queue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Queue
{
    class Queue<T>
    {
        public Queue()
        {
            head = tail = null;
        }

        public void enqueueu(T datum)
        {
            if(size < 1)
            {
                head = tail = new Node<T>(datum);
            }
            else
            {
                Node<T> newNode = new Node<T>(datum);
                tail.next = newNode;
                newNode.prev = tail;
                tail = newNode;
            }
            size++;
        }

        public void dequeue()
        {
            if (size < 1)
            {
                
                Console.WriteLine("there is no Nodes");
                return;
            }
            //T tmp = head.data;
            head = head.next;
            size--;
            if (size < 1) tail = head;
            else head.prev = null;
            //return tmp;
        }

        public override string ToString()
        {
            StringBuilder str = new StringBuilder();
            str.Append("Queue: [ ");
            cur = head;
            for (int i = 0; i < size; i++)
            {
                str.Append(cur.data);
                if (i < size - 1)
                {
                    str.Append(", ");
                    cur = cur.next;
                }
            }
            return str.Append(" ]").ToString();
        }

        public int size { get; private set; } = 0;
        private Node<T> head, tail, cur;
    }
}